src/subsystems.jl

```diff
@@ -0,0 +1,283 @@
+        end
+    end
+end
+
+function handle_component_addition_to_subsystem!(
+    sys::System,
+    subsystem_name::AbstractString,
+    component::RegulationDevice,
+)
+    if !is_assigned_to_subsystem(sys, component.device, subsystem_name)
+        add_component_to_subsystem!(sys, subsystem_name, component.device)
+    end
+end
+
+function handle_component_removal_from_subsystem!(
+    sys::System,
+    subsystem_name::AbstractString,
+    component::RegulationDevice,
+)
+    if is_assigned_to_subsystem(sys, component.device, subsystem_name)
+        remove_component_from_subsystem!(sys, subsystem_name, component.device)
+    end
+end
+
+"""
+Return a Generator of all components in the subsystem.
+
+Throws ArgumentError if the subsystem name is not stored.
+"""
+get_subsystem_components(sys::System, subsystem_name::AbstractString) =
+    IS.get_subsystem_components(sys.data, subsystem_name)
+
+"""
+Remove a component from a subsystem.
+
+Throws ArgumentError if the subsystem name or component is not stored.
+"""
+function remove_component_from_subsystem!(
+    sys::System,
+    subsystem_name::AbstractString,
+    component::Component,
+)
+    IS.remove_component_from_subsystem!(sys.data, subsystem_name, component)
+    handle_component_removal_from_subsystem!(sys, subsystem_name, component)
+    return
+end
+
+remove_component_from_subsystems!(
+    sys::System,
+    component::Component,
+) = remove_component_from_subsystems!(sys.data, component)
+
+"""
+Return true if the component is in the subsystem.
+"""
+has_component(
+    sys::System,
+    subsystem_name::AbstractString,
+    component::Component,
+) = IS.has_component(sys.data, subsystem_name, component)
+
+"""
+Return a Vector of subsystem names that contain the component.
+"""
+get_assigned_subsystems(
+    sys::System,
+    component::Component,
+) = IS.get_assigned_subsystems(sys.data, component)
+
+"""
+Return true if the component is assigned to any subsystems.
+"""
+is_assigned_to_subsystem(sys::System, component::Component) =
+    IS.is_assigned_to_subsystem(sys.data, component)
+
+"""
+Return true if the component is assigned to the subsystem.
+"""
+is_assigned_to_subsystem(
+    sys::System,
+    component::Component,
+    subsystem_name::AbstractString,
+) = IS.is_assigned_to_subsystem(sys.data, component, subsystem_name)
+
+# Utility function, not for export
+get_component_uuids(sys::System, subsystem_name::AbstractString) =
+    IS.get_component_uuids(sys.data, subsystem_name)
+
+function check_subsystems(sys::System, component::Component)
+    _check_arc_consistency(sys, component)
+    _check_branch_consistency(sys, component)
+    _check_device_service_consistency(sys, component)
+    _check_subcomponent_consistency(sys, component)
+    _check_topological_consistency(sys, component)
+    return
+end
+
+function _check_num_subsystems(sys::System)
+    num_buses = length(sys.bus_numbers)
+    if num_buses == get_num_subsystems(sys)
```

Comment:
Is it not necessary to use =< instead of ==?

Reply:
Technically, the code is correct because this gets called at every addition of a subsystem. But your point is good and I will change it.